pkgs/tools/graphics/realcugan-ncnn-vulkan/default.nix

```diff
@@ -0,0 +1,57 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, vulkan-headers
+, vulkan-loader
+, glslang
+, libgcc
+, libwebp
+, ncnn
+}:
+
+stdenv.mkDerivation rec {
+  pname = "realcugan-ncnn-vulkan";
+  version = "20220318";
+
+  src = fetchFromGitHub {
+    owner = "nihui";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-pmh73rBRLRpHy0uNgiAM9Y04GGEhEbR5M13oTf1kOYI=";
+  };
+  sourceRoot = "source/src";
+
+  patches = [
+    # Fixes missing target SPIRV-Tools-opt
+    ./cmakelists.patch
+    # Load models from $out/share/models instead of $out/bin
```

Comment:
I think it'd be better to install the executable  to `$out/opt` instead, and symlink the models to the same directory. Then you could create a symlink from the binary in `$out/opt` to `$out/bin`. 

Just to avoid maintaining another patch.

Reply:
`$out/opt` or `$out/opt/models`?

Reply:
According to the patch used now I think you can't use a seperate `models` folder, so `$out/opt`.

Reply:
Using `$out/opt` feels a little hacky. I couldn’t find anything else in nixpkgs that did that.

The following approach is a bit more work. It updates the patches to define and use `WAIFU2X_MODELDIR` and then sets it in `cmake` based on `CMAKE_INSTALL_DATAROOTDIR`. I then include `"-DCMAKE_INSTALL_DATAROOTDIR=${placeholder "out"}/share"` in the `cmakeFlags` in the derivation. This approach might be something upstream would accept as a patch, which would reduce the maintenance burden if they did.

```patch
diff --git a/main.cpp b/main.cpp
index fa10e6f..77f5d8f 100644
--- a/main.cpp
+++ b/main.cpp
@@ -470,7 +470,11 @@ int main(int argc, char** argv)
     int noise = 0;
     int scale = 2;
     std::vector<int> tilesize;
+#ifdef WAIFU2X_MODELDIR
+    path_t model = PATHSTR(WAIFU2X_MODELDIR);
+#else
     path_t model = PATHSTR("models-cunet");
+#endif
     std::vector<int> gpuid;
     int jobs_load = 1;
     std::vector<int> jobs_proc;
```

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6b73e42..95a7efd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -68,6 +68,8 @@ if(USE_SYSTEM_NCNN)
             include("${GLSLANG_TARGET_DIR}/HLSLTargets.cmake")
         endif()
         include("${GLSLANG_TARGET_DIR}/glslangTargets.cmake")
+        include("${GLSLANG_TARGET_DIR}/SPIRV-Tools/SPIRV-ToolsTarget.cmake")
+        include("${GLSLANG_TARGET_DIR}/SPIRV-Tools-opt/SPIRV-Tools-optTargets.cmake")
         include("${GLSLANG_TARGET_DIR}/SPIRVTargets.cmake")
 
         if (NOT TARGET glslang OR NOT TARGET SPIRV)
@@ -242,6 +244,10 @@
 waifu2x_add_shader(waifu2x_preproc_tta.comp)
 waifu2x_add_shader(waifu2x_postproc_tta.comp)

+if(DEFINED CMAKE_INSTALL_DATAROOTDIR)
+  add_compile_definitions(WAIFU2X_MODELDIR="${CMAKE_INSTALL_DATAROOTDIR}/models/models-cunet")
+endif()
+
 add_custom_target(generate-spirv DEPENDS ${SHADER_SPV_HEX_FILES})

 add_executable(waifu2x-ncnn-vulkan main.cpp waifu2x.cpp)
```

Reply:
I do agree it's a bit hackish, although I'm not sure of a better solution without patching the source. I'd say a comment explaining the workaround would suffice. Maybe `$out/lib` would be a better directory than `$out/opt`?

As you said though, I definitely do think upstreaming a proper patch would be the best move here. If a PR were to be made we could just `fetchpatch` it until its merged.

Reply:
My instinct was to suggest `$out/libexec`, but that’s not good either. Maybe `$out/opt` is the least bad solution considering that it’s mixing data with binaries. I’d say go with that for now.